### ISSUE TITLE
Too many queries to test if address_id exists

### ISSUE THREAD

**Issue by lmeyer1** [PrestaShop/PrestaShop#26754]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

When there is a product in the cart and I'm not logged in, Prestashop is repeating this query 30 times

    SELECT SQL_NO_CACHE `id_address`  FROM ps_address a  WHERE a.`id_address` = 0 LIMIT 1

Additionally the logic is suspect, because the query does not check if the address is marked as deleted or if it is active. This would be necessary, since the addresses are soft-deleted in the address table.

The screenshot are from Profiling mode on a product page. Checking the individual queries in Stopwatch SQL shows that all queries are exactly the same.

![2021-11-25_161510](https://user-images.githubusercontent.com/7520695/143466190-ec24be8e-1271-4f87-b7ba-304fa80108d8.jpg)

### Expected behavior

Never check if address with `id_address = 0` exists. It must never exist.
Check only once if an address exists, remember the answer.
Even better: good execution logic would test only once for the existence of the address saved in the cart.

To-do:

- [x] #28321

- [ ]  Investigate the core of the issue to understand: why we are doing so much queries (related to how we store the adress in the cache of the cart) ?

### Steps to reproduce

1. Enable Profiling mode
3. Add a product to your cart
4. Scroll down to query doubles, you'll notice that there is 30x the same query on the table `address`.

### PrestaShop version(s) where the bug happened

1.7.8.0

### PHP version(s) where the bug happened

7.3

**Comment by HanaRebaiQA** [comment]
Hello @lmeyer1 

Thanks for reporting the issue!
I reproduce the issue with PS1780, see the attached screenshot below:

![image](https://user-images.githubusercontent.com/92912932/143572846-43d0d099-d031-464d-b60d-ae6c5a21affe.png)

It's a regression cause on PS1778 the request is not repeated. See the attached screenshot below : 

![image](https://user-images.githubusercontent.com/92912932/143573395-e2b6fde0-50a2-481a-ae73-b580bfc856cf.png)

I'll add this to the debug backlog so that it's fixed.
Thanks for your PR :rocket: 

Thanks!

**Comment by Dar-us-Salam** [comment]
Hi, I have been experiencing the same issue even in PS 1.7.8.1 and thought it was a third-party module which was causing the issue until I searched and found this page. My shop has large address database so the Doubles are very high in the 1600-2300 range and is occurring on almost all the shop pages when customer is logged in.

Here are a few screen shots:
* Address Query: https://prnt.sc/20y2hei
* Doubles on the product page: https://prnt.sc/20y2j8j -- Secondary issue: The whole style tag shows up instead of just the XXs in blue. I don't know if it is PS or something with my install. Can just add XX in future instead of coloring them blue.
* Query & Doubles on the main page: https://prnt.sc/20y2k1r

Website address: https://dar-us-salam.com/
Hopefully it will get fixed soon. Let me know if anything further is needed.

**Comment by lmeyer1** [comment]
@Dar-us-Salam 
> Address query

There are about 45 x the same query per article in the cart when viewing the cart page, and about 30 x on a product page... Did you have that many articles in your cart?

> Display of *Doubles* in Profiling mode

See #26718

@MatShir, @eternoendless, @jolelievre
Look at https://prnt.sc/20y2j8j from comment above: 2238x the same query looks really crazy !

**Comment by Dar-us-Salam** [comment]
I only had 5 items in the basket (qty 1 of each) when I did Profiling and took the screenshots so that shouldn't be an issue.

**Comment by augustomurri** [comment]
Hi, i have experiencing same issue on PS 1.7.8.5 .. disaster :(

Anyone can help me? I have disabled all the modules

![image](https://user-images.githubusercontent.com/16515870/172903615-49d05d2f-1922-42f9-914f-1960a1ce25ad.png)

**Comment by Progi1984** [comment]
Fixed by #28232

**Comment by Hlavtox** [comment]
@Progi1984 @matks Not fixed, why closing?

**Comment by augustomurri** [comment]
The fixed suggested in [#28232](https://github.com/PrestaShop/PrestaShop/pull/28232) add a caching mechanism in _class/Address_ and _classes/Cart.php_ but don't consider the method **getProductAddressId** in _classes/Cart.php_ wich also use AddressExists method

```
    public function getProductAddressId($product)
    {
        $taxAddressType = $this->configuration->get('PS_TAX_ADDRESS_TYPE');
        if ($taxAddressType == 'id_address_invoice') {
            $addressId = (int) $this->id_address_invoice;
        } else {
            $addressId = (int) $product['id_address_delivery'];
        }

        // Get delivery address of the product from the cart
        /*** FIX HERE ***/
        if (!$this->addressFactory->addressExists($addressId)) {
            $addressId = null;
        }
        /*** FIX HERE ***/

        return $addressId;
    }
```
Commenting the lines suggested above the problem does not happen on my project

In my case ALL the pages make hundred and hundred of duplicated queries on ps_address

You can the the problem [here](https://nuovo.tsunaminutrition.it/) where Profiling mode is active **but you need to register or social login AND some products to basket** (only logged in user suffer the bug)

We are blocked and cannot go online with the project until fixed :(

**Comment by jkulikowski** [comment]
I guess we won't get any update on that? Still a lot of people use 1.7 and struggle with this issue

**Comment by jbgomond** [comment]
Hello. Do we have an update on that issue ? In my Prestashop, this query is called over 6000 times on certain admin pages ...

**Comment by boo-code** [comment]
The part that made this Major is fixed. The second half is still true, and the project has annotated it in the code rather than changed it.

**The repeated query.** `Address::addressExists()` on 9.2.x starts with:

```php
if ((int) $id_address <= 0) {
    return false;
}

if ($useCache && isset(static::$addressExists[$id_address])) {
    return static::$addressExists[$id_address];
}
```

so `WHERE a.id_address = 0` is never sent at all, and a repeated lookup for a real id is answered from the static cache. All three callers on the cart path pass `true` for the cache: `Cart.php:990`, `:2570` and `:3683`. That came from #28780 by @eternoendless, merged in July 2022.

**The soft-delete point.** Still accurate: the query does not look at `deleted`. It is a known caveat rather than an oversight, and it is written down in two places:

```php
/** @var array<int, bool> Store if an adress ID exists. Please note that soft-deleted address also belongs to this cache. */
protected static $addressExists = [];
```

and, where the cache is warmed on save:

```php
// Update the cache
// This is probably not correct, because it should be true only if the address is NOT flagged as deleted
static::$addressExists[$this->id] = true;
```

So somebody looked at exactly this and left the behaviour alone deliberately. Changing it now would make `addressExists()` return false for a soft-deleted address, and the callers above use it to decide whether a cart's delivery or invoice address is usable, so an order placed against an address the customer later deleted would start behaving differently. That is a decision rather than a fix, and worth its own issue if it should change.

Suggest closing this one on the performance grounds it was raised for, and opening the soft-delete question separately if it still matters.
